src/csharp/Grpc.Core/Channel.cs

```diff
@@ -152,8 +152,11 @@ public Task WaitForStateChangedAsync(ChannelState lastObservedState, DateTime? d
                 "Shutdown is a terminal state. No further state changes can occur.");
             var tcs = new TaskCompletionSource<object>();
             var deadlineTimespec = deadline.HasValue ? Timespec.FromDateTime(deadline.Value) : Timespec.InfFuture;
-            // pass "tcs" as "state" for WatchConnectivityStateHandler.
-            handle.WatchConnectivityState(lastObservedState, deadlineTimespec, completionQueue, WatchConnectivityStateHandler, tcs);
+            lock (myLock)
+            {
+                // pass "tcs" as "state" for WatchConnectivityStateHandler.
+                handle.WatchConnectivityState(lastObservedState, deadlineTimespec, completionQueue, WatchConnectivityStateHandler, tcs);
```

Comment:
I might be unclear on where the race is; I'm not sure why destroying of the channel doesn't force that loop to exit.

Reply:
The loop is actually terminating just fine, but there's a race between invoking, but the race between grpc_channel_destroy and grpc_watch_connectivity_state has the effect of not emitting a completion from completion queue, so the tests fails with assertion that there's an unfinished completion in completion registry (and in cq) - because we started the watch_connectivity_state  but we never receive its completion.